File: dotnet/Multiplayer/Matchmaking/Matchmaking.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Michitai.Multiplayer.Matchmaking
{
    /// <summary>
    /// Provides methods for matchmaking lobby management including creation,
    /// joining, player management, and game session initiation.
    /// </summary>
    public class Matchmaking
    {
        /// <summary>
        /// Retrieves a list of available matchmaking lobbies.
        /// </summary>
        /// <typeparam name="T">The type to deserialize lobby rules into.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="search">Optional search term to filter lobbies.</param>
        /// <param name="limit">Optional limit on the number of results.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response containing the list of matchmaking lobbies.</returns>
        public static Task<MatchmakingListResponse<T>> GetMatchmakingLobbiesAsync<T>(Client client, string? search = null, int? limit = null, CancellationToken ct = default) where T : class, new()
            => client.Send<MatchmakingListResponse<T>>(HttpMethod.Post, client.Url(Endpoints.MatchmakingList), new MatchmakingListRequest { Search = search, Limit = limit }, ct);

        /// <summary>
        /// Creates a new matchmaking lobby with specified configuration.
        /// </summary>
        /// <typeparam name="TPlayerData">The type of player data for the host.</typeparam>
        /// <typeparam name="TRules">The type of game rules for the lobby.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The host's player token.</param>
        /// <param name="matchmakingName">The name of the matchmaking lobby.</param>
        /// <param name="maxPlayers">Maximum number of players allowed (default: 4).</param>
        /// <param name="strictFull">Whether the lobby must be full to start (default: false).</param>
        /// <param name="hostSwitch">Whether host switching is allowed (default: false).</param>
        /// <param name="canLeaveRoom">Whether players can leave the resulting room (default: false).</param>
        /// <param name="realtimeRoom">Whether the resulting room supports realtime communication (default: false).</param>
        /// <param name="password">Optional password for the lobby.</param>
        /// <param name="playerData">Optional player data for the host.</param>
        /// <param name="rules">Optional game rules for the lobby.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response containing the matchmaking lobby ID.</returns>
        public static Task<MatchmakingCreateResponse> CreateMatchmakingLobbyAsync<TPlayerData, TRules>(Client client, string playerToken, string matchmakingName, int maxPlayers = 4, bool strictFull = false,
            bool hostSwitch = false, bool canLeaveRoom = false, bool realtimeRoom = false, string? password = null, TPlayerData? playerData = null, TRules? rules = null, CancellationToken ct = default) where TPlayerData : class, new() where TRules : class, new()
            => client.Send<MatchmakingCreateResponse>(HttpMethod.Post, client.Url(Endpoints.MatchmakingCreate, $"&player_token={playerToken}"),
                new MatchmakingCreateRequest<TPlayerData, TRules>(matchmakingName, maxPlayers, strictFull, false, hostSwitch, canLeaveRoom, realtimeRoom, password, playerData, rules), ct);

        /// <summary>
        /// Gets the current status of the player's matchmaking lobby.
        /// </summary>
        /// <typeparam name="T">The type to deserialize lobby rules into.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response containing the current lobby status and information.</returns>
        public static Task<MatchmakingCurrentResponse<T>> GetCurrentMatchmakingStatusAsync<T>(Client client, string playerToken, CancellationToken ct = default) where T : class, new()
            => client.Send<MatchmakingCurrentResponse<T>>(HttpMethod.Get, client.Url(Endpoints.MatchmakingCurrent, $"&player_token={playerToken}"), null, ct);

        /// <summary>
        /// Joins a matchmaking lobby directly (without approval).
        /// Only works if the lobby doesn't require host approval.
        /// </summary>
        /// <typeparam name="T">The type of player data.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="matchmakingId">The ID of the matchmaking lobby to join.</param>
        /// <param name="playerData">Optional player data to include when joining.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response confirming the player joined the lobby.</returns>
        public static Task<MatchmakingDirectJoinResponse> JoinMatchmakingDirectlyAsync<T>(Client client, string playerToken, string matchmakingId, T? playerData = null, CancellationToken ct = default) where T : class, new()
            => client.Send<MatchmakingDirectJoinResponse>(HttpMethod.Post, client.Url(string.Format(Endpoints.MatchmakingJoin, matchmakingId), $"&player_token={playerToken}"), playerData, ct);

        /// <summary>
        /// Leaves the current matchmaking lobby.
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response confirming the player left the lobby.</returns>
        public static Task<MatchmakingLeaveResponse> LeaveMatchmakingAsync(Client client, string playerToken, CancellationToken ct = default)
            => client.Send<MatchmakingLeaveResponse>(HttpMethod.Post, client.Url(Endpoints.MatchmakingLeave, $"&player_token={playerToken}"), null, ct);

        /// <summary>
        /// Gets the list of players in the current matchmaking lobby.
        /// </summary>
        /// <typeparam name="T">The type to deserialize player data into.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response containing the list of players in the lobby.</returns>
        public static Task<MatchmakingPlayersResponse<T>> GetMatchmakingPlayersAsync<T>(Client client, string playerToken, CancellationToken ct = default) where T : class, new()
            => client.Send<MatchmakingPlayersResponse<T>>(HttpMethod.Get, client.Url(Endpoints.MatchmakingPlayers, $"&player_token={playerToken}"), null, ct);

        /// <summary>
        /// Sends a heartbeat to maintain the player's presence in the matchmaking lobby.
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response confirming the heartbeat was received.</returns>
        public static Task<MatchmakingHeartbeatResponse> SendMatchmakingHeartbeatAsync(Client client, string playerToken, CancellationToken ct = default)
            => client.Send<MatchmakingHeartbeatResponse>(HttpMethod.Post, client.Url(Endpoints.MatchmakingHeartbeat, $"&player_token={playerToken}"), null, ct);

        /// <summary>
        /// Removes the current matchmaking lobby (host only).
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The host's authentication token.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response confirming the lobby was removed.</returns>
        public static Task<MatchmakingRemoveResponse> RemoveMatchmakingLobbyAsync(Client client, string playerToken, CancellationToken ct = default)
            => client.Send<MatchmakingRemoveResponse>(HttpMethod.Post, client.Url(Endpoints.MatchmakingRemove, $"&player_token={playerToken}"), null, ct);

        /// <summary>
        /// Starts the game from the current matchmaking lobby and creates a game room.
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The host's authentication token.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response containing the created room ID.</returns>
        public static Task<MatchmakingStartResponse> StartGameFromMatchmakingAsync(Client client, string playerToken, CancellationToken ct = default)
            => client.Send<MatchmakingStartResponse>(HttpMethod.Post, client.Url(Endpoints.MatchmakingStart, $"&player_token={playerToken}"), null, ct);

        /// <summary>
        /// Stops the current matchmaking lobby (host only).
        /// Cannot be called after the game has started.
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The host's authentication token.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Success response confirming the lobby was stopped.</returns>
        public static Task<SuccessResponse> StopMatchmakingAsync(Client client, string playerToken, CancellationToken ct = default)
            => client.Send<SuccessResponse>(HttpMethod.Post, client.Url(Endpoints.MatchmakingStop, $"&player_token={playerToken}"), null, ct);

        /// <summary>
        /// Kicks a player from the matchmaking lobby (host only).
        /// Cannot kick players after the game has started.
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The host's authentication token.</param>
        /// <param name="playerId">The ID of the player to kick.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response confirming the player was kicked.</returns>
        public static Task<MatchmakingKickResponse> KickPlayerAsync(Client client, string playerToken, int playerId, CancellationToken ct = default)
            => client.Send<MatchmakingKickResponse>(HttpMethod.Post, client.Url(Endpoints.MatchmakingKick, $"&player_token={playerToken}"), new MatchmakingKickRequest { Player_id = playerId }, ct);

        /// <summary>
        /// Updates the password for the matchmaking lobby (host only).
        /// Cannot change password after the game has started.
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The host's authentication token.</param>
        /// <param name="password">New password, or null to remove the password.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Success response confirming the password was updated.</returns>
        public static Task<SuccessResponse> UpdateMatchmakingPasswordAsync(Client client, string playerToken, string? password = null, CancellationToken ct = default)
            => client.Send<SuccessResponse>(HttpMethod.Post, client.Url(Endpoints.MatchmakingPassword, $"&player_token={playerToken}"), new MatchmakingPasswordUpdateRequest(password), ct);
    }
}

File: python/multiplayer/matchmaking/matchmaking.py
"""Matchmaking lobby management: create, join, players, start, admin ops."""

from __future__ import annotations

import dataclasses
from typing import Any, Optional, Type, TypeVar

from .._serialization import from_dict
from ..client import Client
from .models import (
    MatchmakingCreateResponse,
    MatchmakingCurrentResponse,
    MatchmakingDirectJoinResponse,
    MatchmakingHeartbeatResponse,
    MatchmakingKickResponse,
    MatchmakingLeaveResponse,
    MatchmakingListResponse,
    MatchmakingPasswordResponse,
    MatchmakingPlayersResponse,
    MatchmakingRemoveResponse,
    MatchmakingStartResponse,
    MatchmakingStopResponse,
)

T = TypeVar("T")

# Endpoints
MATCHMAKING_LIST = "matchmaking.php/list"
MATCHMAKING_CREATE = "matchmaking.php/create"
MATCHMAKING_CURRENT = "matchmaking.php/current"
MATCHMAKING_JOIN = "matchmaking.php/{matchmaking_id}/join"
MATCHMAKING_LEAVE = "matchmaking.php/leave"
MATCHMAKING_PLAYERS = "matchmaking.php/players"
MATCHMAKING_HEARTBEAT = "matchmaking.php/heartbeat"
MATCHMAKING_REMOVE = "matchmaking.php/remove"
MATCHMAKING_START = "matchmaking.php/start"
MATCHMAKING_STOP = "matchmaking.php/stop"
MATCHMAKING_KICK = "matchmaking.php/kick"
MATCHMAKING_PASSWORD = "matchmaking.php/password"


class Matchmaking:
    """Matchmaking lobby management."""

    def __init__(self, client: Client) -> None:
        self._client = client

    def list(
        self,
        search: Optional[str] = None,
        limit: Optional[int] = None,
        rules_cls: Optional[Type[T]] = None,
    ) -> MatchmakingListResponse:
        """List available matchmaking lobbies.

        :param rules_cls: Optional dataclass type for each lobby's ``rules``.
        """
        response = self._client.send(
            "POST",
            self._client.url(MATCHMAKING_LIST),
            {"search": search, "limit": limit},
            MatchmakingListResponse,
        )
        _apply_type(response.lobbies, "rules", rules_cls)
        return response

    def create(
        self,
        player_token: str,
        matchmaking_name: str,
        max_players: int = 4,
        strict_full: bool = False,
        host_switch: bool = False,
        can_leave_room: bool = False,
        realtime_room: bool = False,
        password: Optional[str] = None,
        player_data: Any = None,
        rules: Any = None,
    ) -> MatchmakingCreateResponse:
        """Create a new matchmaking lobby (direct join, no approval needed).

        To require host approval for joins, use
        :meth:`client.matchmaking_requests.create` with ``join_by_requests=True``.
        """
        return self._client.send(
            "POST",
            self._client.player_url(MATCHMAKING_CREATE, player_token),
            {
                "matchmaking_name": matchmaking_name,
                "max_players": max_players,
                "strict_full": strict_full,
                "join_by_requests": False,
                "host_switch": host_switch,
                "can_leave_room": can_leave_room,
                "realtime_room": realtime_room,
                "password": password,
                "player_data": player_data,
                "rules": rules,
            },
            MatchmakingCreateResponse,
        )

    def get_current(
        self,
        player_token: str,
        rules_cls: Optional[Type[T]] = None,
    ) -> MatchmakingCurrentResponse:
        """Get the status of the player's current matchmaking lobby.

        :param rules_cls: Optional dataclass type for the lobby's ``rules``.
        """
        response = self._client.send(
            "GET",
            self._client.player_url(MATCHMAKING_CURRENT, player_token),
            None,
            MatchmakingCurrentResponse,
        )
        if rules_cls is not None and response.matchmaking is not None:
            response.matchmaking.rules = _convert(
                response.matchmaking.rules, rules_cls
            )
        return response

    def join(
        self,
        player_token: str,
        matchmaking_id: str,
        player_data: Any = None,
    ) -> MatchmakingDirectJoinResponse:
        """Join a matchmaking lobby directly (no host approval)."""
        return self._client.send(
            "POST",
            self._client.player_url(
                MATCHMAKING_JOIN.format(matchmaking_id=matchmaking_id), player_token
            ),
            player_data,
            MatchmakingDirectJoinResponse,
        )

    def leave(self, player_token: str) -> MatchmakingLeaveResponse:
        """Leave the current matchmaking lobby."""
        return self._client.send(
            "POST",
            self._client.player_url(MATCHMAKING_LEAVE, player_token),
            None,
            MatchmakingLeaveResponse,
        )

    def get_players(
        self,
        player_token: str,
        data_cls: Optional[Type[T]] = None,
    ) -> MatchmakingPlayersResponse:
        """Get the list of players in the current lobby.

        :param data_cls: Optional dataclass type for each player's ``player_data``.
        """
        response = self._client.send(
            "GET",
            self._client.player_url(MATCHMAKING_PLAYERS, player_token),
            None,
            MatchmakingPlayersResponse,
        )
        _apply_type(response.players, "player_data", data_cls)
        return response

    def heartbeat(self, player_token: str) -> MatchmakingHeartbeatResponse:
        """Send a heartbeat to maintain presence in the lobby."""
        return self._client.send(
            "POST",
            self._client.player_url(MATCHMAKING_HEARTBEAT, player_token),
            None,
            MatchmakingHeartbeatResponse,
        )

    def remove(self, player_token: str) -> MatchmakingRemoveResponse:
        """Remove the current matchmaking lobby (host only)."""
        return self._client.send(
            "POST",
            self._client.player_url(MATCHMAKING_REMOVE, player_token),
            None,
            MatchmakingRemoveResponse,
        )

    def start(self, player_token: str) -> MatchmakingStartResponse:
        """Start the game: creates a game room and transfers all players."""
        return self._client.send(
            "POST",
            self._client.player_url(MATCHMAKING_START, player_token),
            None,
            MatchmakingStartResponse,
        )

    def stop(self, player_token: str) -> MatchmakingStopResponse:
        """Stop the current lobby (host only, before the game started)."""
        return self._client.send(
            "POST",
            self._client.player_url(MATCHMAKING_STOP, player_token),
            None,
            MatchmakingStopResponse,
        )

    def kick(self, player_token: str, player_id: int) -> MatchmakingKickResponse:
        """Kick a player from the lobby (host only, before the game started)."""
        return self._client.send(
            "POST",
            self._client.player_url(MATCHMAKING_KICK, player_token),
            {"player_id": player_id},
            MatchmakingKickResponse,
        )

    def update_password(
        self,
        player_token: str,
        password: Optional[str] = None,
    ) -> MatchmakingPasswordResponse:
        """Update the lobby password (host only, before the game started)."""
        return self._client.send(
            "POST",
            self._client.player_url(MATCHMAKING_PASSWORD, player_token),
            {"password": password},
            MatchmakingPasswordResponse,
        )


def _convert(value: Any, cls: Type[T]) -> Any:
    if dataclasses.is_dataclass(cls) and isinstance(value, dict):
        return from_dict(cls, value)
    return value


def _apply_type(items: Any, attr: str, cls: Optional[Type[T]]) -> None:
    if cls is None:
        return
    for item in items:
        setattr(item, attr, _convert(getattr(item, attr), cls))
